File: DataStructures/com/graphs/problems/RottenOranges.java
package com.graphs.problems;

import java.util.LinkedList;
import java.util.Queue;

public class RottenOranges {

    public static int rottingOranges( int[][] grid ) {

        if( grid == null || grid.length == 0 ) return 0;
        Queue< int[] > rotten = new LinkedList<>();
        int freshOranges = 0;

        for( int i = 0; i < grid.length; i++ ) {
             for( int j = 0; j < grid[i].length; j++ ) {
                  if( grid[i][j] == 2 ) {
                      rotten.offer( new int[]{i, j});
                  } else if( grid[i][j] == 1 ) {
                      freshOranges++;
                  }
             }
        }

       int[][] directions = {{ -1, 0}, { 1, 0}, { 0, -1}, {0, 1}};
        int minute = 0;
        while ( !rotten.isEmpty() ) {

            int size = rotten.size();
            // Process all rotten oranges in the current minute
            for ( int i = 0; i < size; i++ ) {
                int[] current = rotten.poll();
                int currentRow = current[0];
                int currentCol = current[1];

                // check all 4-directional neighbours
                for( int[] dir: directions ) {
                    int newRow = currentRow + dir[0];
                    int newCol = currentCol + dir[1];

                    // if the neighbour is freshOrange make it rotten and add it to the queue
                    if( newRow >= 0 && newRow < grid.length
                        && newCol >= 0 && newCol < grid[0].length
                            && grid[newRow][newCol] == 1 ) {
                        grid[newRow][newCol] = 2;
                        rotten.offer( new int[]{newRow, newCol});
                        freshOranges--;
                    }
                    if( freshOranges == 0 ) {
                        return minute + 1;
                    }

                }
            }
          minute++;

        }

        return -1;

    }



    public static void main(String[] args) {

        int[][] grid = {{1, 2, 1, 1},
                        {0, 1, 1, 2 },
                        {2, 1, 0, 1 },
                        {0, 1, 0, 1 } };

        System.out.println( rottingOranges(grid ));
    }
}

File: TS/Graphs/src/easy/rottenOranges.ts


function rottingOranges( grid: number[][] ) {
   
   if( grid === null || grid.length === 0 ) {
    return -1;
   }

   const rotten: number[][] = [];
   let freshCount = 0;

   for( let row = 0; row < grid.length; row++ ) {

      for( let col = 0; col < grid[row].length; col++ ) {
         
        if( grid[row][col] === 2 ) {
           rotten.push( [row, col ]);
        } else if( grid[row][col] === 1 ) {
           freshCount++;
        }
      }
   }
   console.log("fresh oranges count ",  freshCount);

   

   const directions: number[][] = [[-1, 0], [1, 0], [0, -1], [0, 1]];
   let minutes = 0;

   while ( rotten.length > 0 ) {
     
     const size = rotten.length;

     for( let i = 0; i < size; i++ ) {
       
       const current = rotten.shift()!;
       const currentRow = current[0];
       const currentCol = current[1];

       for( const dir of directions ) {
          
         const newRow = currentRow + dir[0];
         const newCol = currentCol + dir[1];

         if( newRow >=0 && newRow < grid.length 
                       && newCol >= 0 && newCol < grid[0].length 
                       && grid[newRow][newCol] === 1  ) {


           grid[newRow][newCol] = 2;
           rotten.push( [newRow, newCol]);
           freshCount--;             
        }
        if( freshCount === 0 ) {
          return minutes + 1;
        }
       }

     }

     minutes++;
   }
   return -1;
}

const orangegrid = [[1, 2, 1, 1],
[0, 1, 1, 2 ],
[2, 1, 0, 1 ],
[0, 1, 0, 1 ] ];

const result = rottingOranges(orangegrid)
console.log("No of minutes to rot all fresh oranges : ", result);
